### PR TITLE
SGE implementation

### LINKED ISSUE #3: Add Grid Engine
It might be valuable to extend this repository with a solution for Grid Engine.  My hope is that this is relatively easy for someone with modest grid engine experience.  Looking at the current solutions for [PBS](https://github.com/dask/dask-jobqueue/blob/master/dask_jobqueue/pbs.py) or [SLURM](https://github.com/dask/dask-jobqueue/blob/master/dask_jobqueue/slurm.py) might be helpful (they're about 100 lines, mostly docstrings)

### PR DESCRIPTION
**Pull request by lesteve**
I revived #6 and added simple tests. I was able to get `test_basic` to pass but not `test_adaptive` yet. 

I'll wait that #25 is merged before working on this further.

Fix #3. Closes #6.

### CODE REVIEW

File: dask_jobqueue/tests/test_sge.py [hunk 1]
```diff
@@ -1,9 +1,70 @@
...
+                assert time() < start + 10
+
+            start = time()
+            while cluster.jobs:
+                sleep(0.100)
+                assert time() < start + 10
+
+
+@pytest.mark.skipif('SGE_ACCOUNT' in os.environ, reason='SGE_ACCOUNT defined')  # noqa: F811
+def test_errors(loop):
```
Comment (jhamman) [inline dask_jobqueue/tests/test_sge.py]: I don't think this test is relevant anymore.

File: dask_jobqueue/tests/test_sge.py [hunk 2]
```diff
@@ -1,9 +1,70 @@
...
+
+            start = time()
+            while len(client.scheduler_info()['workers']) > 0:
+                sleep(0.100)
+                assert time() < start + 10
+
+            start = time()
+            while cluster.jobs:
+                sleep(0.100)
+                assert time() < start + 10
```
Comment (jhamman) [inline dask_jobqueue/tests/test_sge.py]: What is this testing exactly?  @mrocklin - given the recent development of `Adaptive`, can you help us flesh out what kind of tests we should be looking at here?

File: dask_jobqueue/tests/test_sge.py (hunk 1 above)
Comment (lesteve) [reply dask_jobqueue/tests/test_sge.py]: Yeah I was wondering about that and I felt the same way.

Comment (mrocklin) [reply dask_jobqueue/tests/test_sge.py]: I think that this test came from a cluster where `project=` was necessary to run, and was often set implicitly with an environment variable (perhaps `PBS_ACCOUNT` on the cluster on which I was working at the time).  If this is not generally true for SGE clusters then I agree that this test should be removed.

File: dask_jobqueue/tests/test_sge.py (hunk 2 above)
Comment (mrocklin) [reply dask_jobqueue/tests/test_sge.py]: I've simplified the test for now.

Comment (mrocklin) [reply dask_jobqueue/tests/test_sge.py]: (I hope you don't mind my pushing to your fork @lesteve )

Comment (lesteve) [reply dask_jobqueue/tests/test_sge.py]: > I've simplified the test for now.

Thanks!

> (I hope you don't mind my pushing to your fork @lesteve )

Not at all, pushing into a PR's fork is a feature I use this feature quite often myself, it is very convenient and saves a lot of back and forth for minor things!

**Comment by lesteve** [comment]
OK I am removing the WIP. A few comments:
* I could not find an equivalent of PBS_ACCOUNT in SGE so I removed the associated code and `test_errors` test
* AFAICT adaptive is currently broken, so I am leaving `test_adaptive` for later.

IMO this should be merged like this since this is already a clear improvement. It is a first implementation of SGE (probably by no means perfect) and allows to test some of the common code in `dask_jobqueue/core.py` across schedulers.

More details about the adaptive problems I saw (maybe related to https://github.com/dask/dask-jobqueue/issues/26 not sure):
According to https://github.com/dask/dask-jobqueue/blob/bdb2c2630d440aad8f7e9a47aeb7cfc2b6776129/dask_jobqueue/core.py#L220-L224, `scale_down` only does something `if isinstance(workers, dict)`. The problem is that workers is a list of `tcp://...` worker addresses. So `scale_down` does nothing i.e. `self.jobs` stays the same so that `scale_up` doesn't see the need to schedule more jobs:
https://github.com/dask/dask-jobqueue/blob/bdb2c2630d440aad8f7e9a47aeb7cfc2b6776129/dask_jobqueue/core.py#L217-L218
In other words your `*Cluster` Python object think he still have enough workers but your scheduling system has no jobs in running state that can process the work.

I guess this is all fixable by looking at the [doc](http://distributed.readthedocs.io/en/latest/adaptive.html#adaptive-class-interface), the `distributed.LocalCluster` implementation and possibly dask_drmaa as well. I may not find time to look at this for a few days so if anyone wants to pick this up, please be my guest!

File: dask_jobqueue/sge.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    #Override class variables
+    submit_command = 'qsub -terse'
+    cancel_command = 'qdel'
+
+    def __init__(self,
+                 queue=None,
+                 project=None,
+                 resource_spec=None,
+                 walltime='0:30:00',
+                 interface=None,
```
Comment (guillaumeeb) [inline dask_jobqueue/sge.py]: I believe interface=None should not be here. It is a keyword from the parent class, and you don't seem to use it.

File: dask_jobqueue/sge.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    Parameters
+    ----------
+    name : str
+        Name of worker jobs. Passed to `$SGE -N` option.
+    queue : str
+        Destination queue for each worker job. Passed to `#$ -q` option.
+    project : str
+        Accounting string associated with each worker job. Passed to
+        `#$ -A` option.
+    threads_per_worker : int
```
Comment (guillaumeeb) [inline dask_jobqueue/sge.py]: For all the parameters inherited from JobQueueCluster, you should use the same mechanism as in PBSCluster or SLURMCluster, with the docstrings module: `%(JobQueueCluster.parameters)s`

File: dask_jobqueue/core.py [hunk 5]
```diff
@@ -153,8 +154,8 @@ def start_workers(self, n=1):
         workers = []
         for _ in range(n):
             with self.job_file() as fn:
-                out = self._call([self.submit_command, fn])
-                job = out.decode().split('.')[0]
+                out = self._call(shlex.split(self.submit_command) + [fn])
```
Comment (guillaumeeb) [inline dask_jobqueue/core.py]: I don't know shlex, is this mandatory? It serves what purpose?

File: dask_jobqueue/sge.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+        self.config = {'name': self.name,
+                       'queue': queue,
+                       'project': project,
+                       'processes': self.worker_processes,
+                       'walltime': walltime,
+                       'resource_spec': resource_spec,}
+        self.job_header = self._header_template % self.config
+
+        logger.debug("Job script: \n %s" % self.job_script())
```
Comment (guillaumeeb) [inline dask_jobqueue/sge.py]: We should probably debug log only the header here, but this is also true for PBS or SLURM, and it is really a minor detail...

**Comment by guillaumeeb** [comment]
And about your other remarks:
* I totally agree to remove the PBS_ACCOUNT part and according test,
* For adaptive, you're completely right about the scale_down method, don't know though if it also breaks the scale_up. If you started with lets say only 1 worker, the scale_up should probably be called at some point, and not the scale_down, so something should happen. But yeah, we should look elsewhere!

File: dask_jobqueue/core.py (hunk 5 above)
Comment (lesteve) [reply dask_jobqueue/core.py]: In SGE, by default qsub returns quite a verbose output, e.g. something like `Your job 56 ("test.sh") has been submitted`. In order for just the job id to be returned, you need to use `qsub -terse`. This is why `submit_cmd = 'qsub -terse'`.

That means that you need to split `submit_cmd`. I think `shlex.split` is the way to do it for sh commands. We could just do `str.split but it may break e.g. if one of the arguments is quoted with a space inside.

File: dask_jobqueue/sge.py (hunk 3 above)
Comment (lesteve) [reply dask_jobqueue/sge.py]: Good catch, done.

File: dask_jobqueue/sge.py (hunk 4 above)
Comment (lesteve) [reply dask_jobqueue/sge.py]: Done.

**Comment by lesteve** [comment]
@guillaumeeb thanks for the review! I tackled your comments, let me know if you have some further comments!

File: dask_jobqueue/sge.py (hunk 4 above)
Comment (guillaumeeb) [reply dask_jobqueue/sge.py]: I believe the sge.py file still misses some import and annotation, e.g. 
https://github.com/dask/dask-jobqueue/blob/2e406ae462b5c82fa679075e45bb2bfa59615610/dask_jobqueue/pbs.py#L5
and
https://github.com/dask/dask-jobqueue/blob/2e406ae462b5c82fa679075e45bb2bfa59615610/dask_jobqueue/pbs.py#L10

You also still have the description of interface kw in the docstring!

Comment (lesteve) [reply dask_jobqueue/sge.py]: Ah right I fixed that, I am not so familiar with the docrep magic!

File: dask_jobqueue/sge.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        Destination queue for each worker job. Passed to `#$ -q` option.
+    project : str
+        Accounting string associated with each worker job. Passed to
+        `#$ -A` option.
+    resource_spec : str
+        Request resources and specify job placement. Passed to `#$ -l`
+        option.
+    walltime : str
+        Walltime for each worker job.
+    interface : str
```
Comment (guillaumeeb) [inline dask_jobqueue/sge.py]: You still have the interface keyword in this docstring, once remove I think we're done :)

Comment (lesteve) [reply dask_jobqueue/sge.py]: Arrggh good point, I just pushed the fix.

File: dask_jobqueue/core.py [hunk 8]
```diff
@@ -153,8 +154,8 @@ def start_workers(self, n=1):
...
+                job = out.decode().split('.')[0].strip()
```
Comment (lesteve) [inline dask_jobqueue/core.py]: And by the way the `.strip()` here is necessary because `qsub -terse` output finishes with a `\n`.

**Comment by jhamman** [comment]
Thanks @lesteve!

**Comment by mrocklin** [comment]
Thanks @lesteve!

**Comment by lesteve** [comment]
Nice to see this merged! I'll try to find some time to fix the adaptive problem.
